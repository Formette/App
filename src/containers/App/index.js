// @flow
import React, { PureComponent } from "react";
import { withRouter } from "react-router-dom";
import { compose } from "react-apollo";
//Components
import { Alert } from "../../components/atoms";
import { Footer } from "../../components/organisms";
import Navbar from "./Navigation";
import MobileNavbar from "./MobileNavigation";
//Context
import { UserContext } from "../../context/UserContext";
//hocs
import { withUser } from "../../hocs";
//Styles
import styled, { withTheme } from "styled-components";
//Pages Navigation
import Router from "../../Router";

class App extends PureComponent {
  state = {
    accountConfirmed: true
  };
  componentWillReceiveProps(nextProps) {
    const { confirmed } = nextProps.user.state.profile;
    this.setState({ accountConfirmed: confirmed });
  }
  render() {
    return (
      <div className={this.props.className}>
        <UserContext.Consumer>
          {context => (
            <Navbar brand="Formette Î²" username={context.state.userName} />
          )}
        </UserContext.Consumer>
        {!this.state.accountConfirmed && (
          <Alert className="alert-warning" role="alert">
            You have not activated your account yet.{" "}
            <a href="#/confirm" className="alert-link">
              Click here to activate.
            </a>
          </Alert>
        )}
        <div className="container-fluid content">
          <Router />
        </div>
        <Footer />
        <MobileNavbar />
      </div>
    );
  }
}

const AppWithStyles = styled(App)`
  margin-top: 100px;
  .content {
    min-height: calc(88vh - 70px);
  }
`;

export default compose(
  withUser,
  withRouter,
  withTheme
)(AppWithStyles);
