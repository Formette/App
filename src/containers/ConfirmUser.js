// @flow
import React from "react";
import { graphql, compose } from "react-apollo";
//Components
import AlertContainer from "react-alert";
import AuthLayout from "../components/organisms/AuthLayout";
import { Input, Button, Link, Icon } from "../components/atoms/index";
import { Error, Graphic } from "../components/molecules/index";
//Styles
import Colors from "../styles/Colors";
//API
import { USER_QUERY } from "../api/Queries";
import {
  USER_CONFIRM_TOKEN_MUTATION,
  USER_RESEND_CONFIRMATION_MUTATION
} from "../api/Mutations";
//Utils
import { getUrlParam, _refreshPage } from "../services/utilities";
import { ALERT_OPTIONS } from "../services/Constants";
import LogRocket from "logrocket";

export class ConfirmUser extends React.PureComponent {
  msg: () => mixed;
  props: {
    userQuery: any,
    confirmEmail: any,
    resendConfirmation: any,
    history: any,
    router: any
  };
  state = {
    confirmToken: "",
    error: false,
    errorMsg: ""
  };
  componentDidMount() {
    if (getUrlParam("token")) {
      this.setState({
        confirmToken: getUrlParam("token")
      });
    }
  }
  showAlert(
    type: string = "success",
    text: string = "Some Text",
    color: string = Colors.green,
    icon: string = "fa-link"
  ) {
    this.msg.show(text, {
      time: 3000,
      type,
      icon: <Icon name={icon} color={color} />
    });
  }
  _onSendConfirmationCode = async () => {
    //verifies the user and saves in the DB
    try {
      const { confirmToken } = this.state;
      await this.props.confirmEmail({
        variables: {
          confirmToken
        }
      });
      //redirects the user to the main page
      this.props.history.push("/");
    } catch (e) {
      LogRocket.error({ SendConfirmationCode: e });
      this.setState({
        error: true,
        errorMsg: "The verification code is invalid."
      });
    }
  };
  _onResendConfirmationCode = async () => {
    //If the code is invalid resend a new code
    try {
      const email = this.props.userQuery.user.email;
      await this.props.resendConfirmation({
        variables: {
          email
        }
      });
      this.showAlert(
        "success",
        "Your code has been resent, check your email.",
        Colors.green,
        "fa-envelope"
      );
    } catch (e) {
      LogRocket.error({ ResendConfirmationCode: e });
      this.setState({
        error: true,
        errorMsg: "The verification code is invalid."
      });
    }
  };
  render() {
    if (this.props.userQuery && this.props.userQuery.loading) {
      return <div>Loading</div>;
    }
    if (this.props.userQuery && this.props.userQuery.error) {
      return (
        <Graphic text="Ups! Something went wrong try again." icon="fa-plug">
          <Button
            className="btn btn-lg btn-primary"
            color={Colors.primary}
            onClick={_refreshPage}
          >
            Try Again
          </Button>
        </Graphic>
      );
    }
    if (this.props.userQuery.user === null) {
      this.props.history.push("/signin");
      return true;
    }
    if (this.props.userQuery.user.confirmed) {
      this.props.history.push("/");
      return true;
    }
    const { confirmToken, error, errorMsg } = this.state;
    return (
      <AuthLayout
        title="Please confirm your email"
        description="We like real people, we need to know if it's not a ghost of the internet."
      >
        <AlertContainer ref={(a) => (this.msg = a)} {...ALERT_OPTIONS} />
        <label htmlFor="confirmToken" className="sr-only">
          Confirmation Code
        </label>
        <Input
          id="confirmToken"
          value={confirmToken}
          onChange={(e) => this.setState({ confirmToken: e.target.value })}
          className="form-control"
          placeholder="Confirmation Code"
          required
          autoFocus
        />

        <Button
          className="btn btn-lg btn-primary btn-block"
          style={{ marginTop: 10, marginBottom: 10 }}
          color={Colors.primary}
          onClick={this._onSendConfirmationCode}
        >
          Confirm Account
        </Button>
        <Link onClick={this._onResendConfirmationCode}>
          Resend confirmation code?
        </Link>
        <Error show={error}>{errorMsg}</Error>
      </AuthLayout>
    );
  }
}

const confirmUserWithData = compose(
  graphql(USER_CONFIRM_TOKEN_MUTATION, { name: "confirmEmail" }),
  graphql(USER_RESEND_CONFIRMATION_MUTATION, { name: "resendConfirmation" }),
  graphql(USER_QUERY, { name: "userQuery" })
);

export default confirmUserWithData(ConfirmUser);
