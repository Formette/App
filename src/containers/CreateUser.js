// @flow
import React from "react";
import { graphql, compose, withApollo } from "react-apollo";
//Components
import AuthLayout from "../components/organisms/AuthLayout";
import { Input, Button, Link } from "../components/atoms/index";
import Error from "../components/molecules/Error";
//Styles
import Colors from "../styles/Colors";
//API
import { USERNAME_VALIDATION_QUERY } from "../api/Queries";
import { SIGIN_USER_MUTATION, CREATE_USER_MUTATION } from "../api/Mutations";
import { userSignIn } from "../api/Functions";
//Utils
import {
  generateToken,
  generateExpiration,
  _formatUsername
} from "../services/utilities";
import LogRocket from "logrocket";

export class CreateUser extends React.PureComponent {
  props: {
    createUser: any,
    signinUser: any,
    history: any,
    client: any,
    router: any
  };
  state = {
    email: "",
    password: "",
    username: "",
    error: false,
    errorMsg: "",
    timeoutUserName: 0,
    timeoutPassword: 0
  };
  _onCreateUser = () => {
    const { email, password, error } = this.state;
    let username = _formatUsername(this.state.username);
    //Verifies if the inputs are empty or not
    if (email && password && username) {
      if (error) {
        return;
      }
      if (this._checkPassword(password)) return;
      //Creates a new user
      this.props
        .createUser({
          variables: {
            email,
            password,
            username,
            confirmToken: generateToken(),
            confirmExpires: generateExpiration()
          }
        })
        .then(() => {
          LogRocket.track("Registered");
          this._onSignIn(email, password);
        })
        .catch(e => {
          LogRocket.error({ CreateUser: e });
          this.setState({
            error: true,
            errorMsg:
              "Ops! User already exists with that information, try again."
          });
        });
    } else {
      this.setState({
        error: true,
        errorMsg: "This form is feeling lonely, needs affection, needs data."
      });
    }
  };
  _onSignIn = async (email: string, password: string) => {
    //logs in the user
    const response = await userSignIn(email, password, this.props.signinUser);
    if (response.status) {
      LogRocket.track("Sign in");
      //checks if the user has the email confirmed
      if (response.confirmed) {
        //sends to the dashboard
        this.props.history.push("/");
      } else {
        this.props.history.push("/confirm");
      }
    } else {
      LogRocket.error({ SignIn: response });
      this.setState({
        error: true,
        errorMsg: "Ops! Something went wrong, try again."
      });
    }
  };
  _onPasswordValidation(password: string) {
    clearTimeout(this.state.timeoutPassword);
    this.setState({
      timeoutPassword: setTimeout(() => {
        this._checkPassword(password);
      }, 500)
    });
  }
  _checkPassword(password: string) {
    if (password.length <= 8) {
      LogRocket.warn(
        "With so much room in the box, you chose this tiny thing. We need more than 8 characters, go we know you can."
      );
      this.setState({
        error: true,
        errorMsg:
          "With so much room in the box, you chose this tiny thing. We need more than 8 characters, go we know you can."
      });
      return true;
    }
    this.setState({ error: false });
  }
  _onUsernameValidation(getUsername: string) {
    clearTimeout(this.state.timeoutUserName);
    this.setState({
      timeoutUserName: setTimeout(() => {
        let username = _formatUsername(getUsername);
        this.props.client
          .query({
            query: USERNAME_VALIDATION_QUERY,
            variables: { username }
          })
          .then(res => {
            if (Object.keys(res.data.allUsers).length !== 0) {
              LogRocket.info(
                "With so much name in this world, you had to choose this one. Try another."
              );
              this.setState({
                error: true,
                errorMsg:
                  "With so much name in this world, you had to choose this one. Try another."
              });
            } else {
              this.setState({ error: false });
            }
          })
          .catch(e => {
            LogRocket.error({ _onUsernameValidation: e });
          });
      }, 500)
    });
  }
  render() {
    const { email, password, username, error, errorMsg } = this.state;
    const { history } = this.props;
    return (
      <AuthLayout description="Seriously? Your forms do not have a home for the data? Do not worry, they will not be without shelter.">
        <label htmlFor="signupUsername" className="sr-only">
          Username
        </label>
        <Input
          id="signupUsername"
          value={username}
          onChange={e => this.setState({ username: e.target.value })}
          onKeyUp={e => this._onUsernameValidation(e.target.value)}
          className="form-control"
          placeholder="Username"
          required
          autoFocus
        />
        <label htmlFor="signupEmail" className="sr-only">
          Email address
        </label>
        <Input
          id="signupEmail"
          type="email"
          value={email}
          onChange={e => this.setState({ email: e.target.value })}
          className="form-control"
          placeholder="Email address"
          required
          autoFocus
        />

        <label htmlFor="signupPassword" className="sr-only">
          Password
        </label>
        <Input
          id="signupPassword"
          type="password"
          value={password}
          onChange={e => this.setState({ password: e.target.value })}
          onKeyUp={e => this._onPasswordValidation(e.target.value)}
          className="form-control"
          placeholder="Password"
          required
          autoFocus
        />

        <Button
          className="btn btn-lg btn-primary btn-block"
          onClick={this._onCreateUser}
          style={{ marginTop: 10, marginBottom: 10 }}
          color={Colors.primary}
        >
          Create Account
        </Button>

        <Link onClick={() => history.push("/signin")}>
          Already have an account? Your forms miss you.{" "}
          <u>
            <strong>Sign In</strong>
          </u>
        </Link>
        <Error show={error}>{errorMsg}</Error>
      </AuthLayout>
    );
  }
}

const CreateUserWithData = compose(
  graphql(CREATE_USER_MUTATION, { name: "createUser" }),
  graphql(SIGIN_USER_MUTATION, { name: "signinUser" })
);

export default withApollo(CreateUserWithData(CreateUser));
