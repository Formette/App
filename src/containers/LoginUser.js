// @flow
import React from 'react';
import { graphql, compose } from 'react-apollo'
//Components
import AuthLayout from '../components/organisms/AuthLayout';
import {Input, Button, Link} from '../components/atoms/index';
import Error from '../components/molecules/Error';
//Styles
import Colors from '../styles/Colors';
//API
import {SIGIN_USER_MUTATION} from '../api/Mutations';
//Utilities
import {_saveUsername, _saveUserId} from '../services/utilities';

class LoginUser extends React.PureComponent {
    props: {
        signinUser: any,
        router: any,
    };
    state = {
        email: "",
        password: "",
        error: false,
        errorMsg: ""
    };
    _onSignIn = _ => {
      const {email, password} = this.state;
        //Verifies if the inputs are empty or not
        if(email && password) {
            //logs in the user
            this.props.signinUser({variables: {email, password}})
                .then((res) => {
                    window.localStorage.setItem('graphcoolToken', res.data.signinUser.token);
                    _saveUsername(res.data.signinUser.user.userName);
                    _saveUserId(res.data.signinUser.user.id);
                }).then(_ => this.props.router.replace('/'))
                .catch((e) => {
                    console.log(e);
                    this.setState({error: true, errorMsg: "Ops! Invalid Email or password."});
                })
        }else{
            this.setState({
                error: true,
                errorMsg: "This form is feeling lonely, needs affection, needs data."
            });
        }
    };
    render(){
        const {email, password, error, errorMsg} = this.state;
        return(
           <AuthLayout description="Welcome back, Come quick! Your forms are waiting for you">
               <label htmlFor="signinEmail" className="sr-only">Email address</label>
                   <Input id="signinEmail"
                          type="email"
                          value={email}
                          onChange={(e) => this.setState({email: e.target.value})}
                          className="form-control"
                          placeholder="Email address"
                          required autoFocus/>

               <label htmlFor="signinPassword" className="sr-only">Password</label>
                   <Input id="signinPassword"
                          type="password"
                          value={password}
                          onChange={(e) => this.setState({password: e.target.value})}
                          className="form-control"
                          placeholder="Password"
                          required autoFocus/>

               <Button className="btn btn-lg btn-primary btn-block"
                       onClick={this._onSignIn}
                       style={{marginTop: 10, marginBottom: 10}}
                       color={Colors.primary}>Sign In</Button>

               <Link href="/signup">Do not have an account yet? Omg is free. <u><strong>Create here!</strong></u></Link>
               <Error show={error}>{errorMsg}</Error>
           </AuthLayout>
        )
    }
}

const LoginUserWithData = compose(
    graphql(SIGIN_USER_MUTATION, {name: 'signinUser'}),
);

export default LoginUserWithData(LoginUser);


