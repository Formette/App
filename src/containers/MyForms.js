// @flow
import React, { Component } from "react";
import { graphql, compose } from "react-apollo";
import * as moment from "moment";
//Components
import { Header, Button } from "../components/atoms/index";
import {
  Placeholder,
  PlaceholderAnimation,
  Card,
  Graphic
} from "../components/molecules/index";
//Utils
import { _refreshPage, _getUserId } from "../services/utilities";
import LogRocket from "logrocket";
//Styles
import Colors from "../styles/Colors";
//API
import { ALL_FORMS_QUERY } from "../api/Queries";

export class MyForms extends Component {
  props: {
    allFormsQuery: any,
    history: any
  };
  state = {
    loading: true,
    loadingTimeout: 0,
    hasError: "",
    data: []
  };
  _LoadingAnimationContent = (type: string = "loading", length: number = 6) => {
    const content = [];
    for (let i = 0; i <= length; i++) {
      if (type !== "loading") {
        content.push(<Placeholder key={i} width={208} height={193} />);
      } else {
        content.push(<PlaceholderAnimation key={i} />);
      }
    }
    return content;
  };
  _loadContent = () => {
    const { allFormses } = this.props.allFormsQuery;
    const formsCount = Object.keys(allFormses).length;
    const content = [];
    let length = 0;
    //checks if the number of forms created from the user
    if (formsCount === 0) {
      return this._LoadingAnimationContent("normal");
    }
    //Checks if the object data is not empty
    if (formsCount >= 0) {
      allFormses.map((res) => {
        content.push(
          <Card
            key={res.id}
            title={res.name}
            date={moment(res.createdAt).format("ll")}
            onClick={() => this.props.history.push(`/form/${res.id}`)}
          />
        );
        return true;
      });
      //for interface
      if (formsCount < 6) {
        length = 6 - Number(formsCount);
      }
      LogRocket.track("Loaded Forms");
      return content.concat(this._LoadingAnimationContent("normal", length));
    }
  };
  _goToNew = () => {
    this.props.history.push("/new");
  };
  render() {
    if (this.props.allFormsQuery && this.props.allFormsQuery.error) {
      return (
        <Graphic text="Ups! Something went wrong try again." icon="fa-plug">
          <Button
            className="btn btn-lg btn-primary"
            color={Colors.primary}
            onClick={_refreshPage}
          >
            Try Again
          </Button>
        </Graphic>
      );
    }
    return (
      <div className="row">
        <div className="col-md-12 col-sm-12 col">
          <Header text="My Forms" />
          <Card
            icon="fa-plus"
            title="New Form"
            date="Create from scratch"
            onClick={this._goToNew}
            new
          />
          {this.props.allFormsQuery.loading ? (
            this._LoadingAnimationContent()
          ) : (
            <div>{this._loadContent()}</div>
          )}
        </div>
      </div>
    );
  }
}

const myFormsWithData = compose(
  graphql(ALL_FORMS_QUERY, {
    name: "allFormsQuery",
    options: (props) => ({
      variables: { userId: _getUserId() }
    })
  })
)(MyForms);

export default myFormsWithData;
