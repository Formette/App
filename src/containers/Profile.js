// @flow
import React, { PureComponent } from "react";
import { graphql, compose, withApollo } from "react-apollo";
//Components
import {
  Header,
  SubTitle,
  Title,
  Input,
  Button,
  Text,
  Link
} from "../components/atoms/index";
import { Graphic, Confirmation } from "../components/molecules/index";
//Styles
import Colors from "../styles/Colors";
//Utils
import {
  _getUsername,
  _logout,
  _saveUsername,
  _refreshPage,
  _formatUsername
} from "../services/utilities";
import LogRocket from "logrocket";
//API
import { USER_QUERY, USERNAME_VALIDATION_QUERY } from "../api/Queries";
import { UPDATE_USER_MUTATION } from "../api/Mutations";

export class Profile extends PureComponent {
  props: {
    userQuery: any,
    updateUser: any,
    client: any,
    router: any,
    updateUsername: any,
    showMessage: () => mixed
  };
  state = {
    username: "",
    error: false,
    errorMsg: "",
    timeoutUserName: 0,
    onConfirmation: false
  };
  componentDidMount() {
    this.setState({ username: _getUsername() });
  }
  _showConfirmation = () => {
    this.setState((prevState) => ({
      onConfirmation: !prevState.onConfirmation
    }));
    LogRocket.track("Opened  modal on change username");
  };
  _updateProfile = async () => {
    const { error } = this.state;
    let username = _formatUsername(this.state.username);
    //hides the confirmation modal
    this.setState(prevState => ({
      onConfirmation: !prevState.onConfirmation
    }));
    //checks is the username is the same as the previous one
    if (this._isTheSameUsername(username)) return;
    //Verifies if the inputs are empty or not
    if (username) {
      if (error) {
        return;
      }
      const userId = this.props.userQuery.user.id;
      //updates the user username and some else info in the DB
      try {
        await this.props.updateUser({
          variables: {
            userId,
            username
          }
        });
        //Shows feedback and updates the localStorage
        _saveUsername(username);
        this.setState({ username });
        //this updates the navbar to the new username
        this.props.updateUsername();
        this.props.showMessage(
          "success",
          "Change made successfully",
          undefined,
          undefined
        );
        LogRocket.info("Change made successfully");
        LogRocket.track("Updated username");
      } catch (e) {
        LogRocket.error({ _updateProfile: e });
        this.props.showMessage(
          "error",
          "Something went wrong, try again...",
          Colors.red,
          "fa-times"
        );
      }
    } else {
      LogRocket.warn(
        "This form is feeling lonely, needs affection, needs data."
      );
      this.setState({
        error: true,
        errorMsg: "This form is feeling lonely, needs affection, needs data."
      });
    }
  };
  _onUsernameValidation(getUsername: string) {
    clearTimeout(this.state.timeoutUserName);
    this.setState({
      timeoutUserName: setTimeout(() => {
        //checks is the username is the same as the previous one
        let username = _formatUsername(getUsername);
        if (this._isTheSameUsername(username)) return;
        this.props.client
          .query({
            query: USERNAME_VALIDATION_QUERY,
            variables: { username }
          })
          .then((res) => {
            if (Object.keys(res.data.allUsers).length !== 0) {
              LogRocket.warn(
                "With so much name in this world, you had to choose this one. Try another."
              );
              this.setState({
                error: true,
                errorMsg:
                  "With so much name in this world, you had to choose this one. Try another."
              });
            } else {
              this.setState({ error: false });
            }
          })
          .catch((e) => {
            LogRocket.error({ _onUsernameValidation: e });
          });
      }, 500)
    });
  }
  _isTheSameUsername(username: string) {
    if (_getUsername() === username) {
      LogRocket.warn(
        "If it's the same as before, what's the point of changing?"
      );
      this.setState({
        error: true,
        errorMsg: "If it's the same as before, what's the point of changing?"
      });
      return true;
    }
  }
  render() {
    if (this.props.userQuery && this.props.userQuery.loading) {
      return <div>Loading</div>;
    }
    if (this.props.userQuery && this.props.userQuery.error) {
      return (
        <Graphic text="Ups! Something went wrong try again." icon="fa-plug">
          <Button
            className="btn btn-lg btn-primary"
            color={Colors.primary}
            onClick={_refreshPage}
          >
            Try Again
          </Button>
        </Graphic>
      );
    }
    const { userName, _formsesMeta } = this.props.userQuery.user;
    const { error, errorMsg, username, onConfirmation } = this.state;
    return (
      <div>
        <Confirmation
          title="Are you sure?"
          description="All your endpoints will be changed to the new username. Do not forget to change in your apps."
          show={onConfirmation}
          onCancel={this._showConfirmation}
          onConfirmation={this._updateProfile}
          onConfirmationText="Confirm"
          onConfirmationColor={Colors.green}
        />
        <div className="row">
          <div className="col-md-12">
            <Header text={`Hey, ${username}`} />
          </div>
        </div>
        <div className="row">
          <div className="col-md-6">
            <form>
              <div className={`form-group ${error ? "has-danger" : ""}`}>
                <SubTitle text="Username:" color={Colors.text.secondary} />
                <Input
                  placeholder="username"
                  defaultValue={userName}
                  onKeyUp={(e) => this._onUsernameValidation(e.target.value)}
                  onChange={(e) =>
                    this.setState({ username: e.target.value, error: false })}
                  className="form-control"
                />
                {error ? <Text text={errorMsg} color={Colors.red} /> : ""}
              </div>
              <Button
                onClick={this._showConfirmation}
                color={Colors.primary}
                className="btn btn-lg btn-primary btn-block"
              >
                Update profile
              </Button>
            </form>
          </div>
          <div className="col-md-6">
            <div className="card" style={{ marginTop: 10 }}>
              <div className="card-body">
                <SubTitle text="Statistics:" color={Colors.text.secondary} />
                <Text
                  text="We are great analysts, here you have your statistics of your forms."
                  color={Colors.text.secondary}
                />
                <Title text={`${_formsesMeta.count} forms created`} />
              </div>
            </div>
            <div className="card" style={{ marginTop: 10 }}>
              <div className="card-body">
                <SubTitle text="Current Plan:" color={Colors.text.secondary} />
                <Text
                  text="More plans soon, as the platform is in beta you have access to all features."
                  color={Colors.text.secondary}
                />

                <h2>
                  <span className="badge badge-dark">All the features</span>
                </h2>
              </div>
            </div>
            <div className="card" style={{ marginTop: 10 }}>
              <div className="card-body">
                <SubTitle text="Settings:" color={Colors.text.secondary} />
                <Text
                  text="This is a dangerous zone, so be careful, here you will find your settings."
                  color={Colors.text.secondary}
                />
                <ul className="list-inline">
                  <li>
                    <Link onClick={_logout}>Log Out</Link>
                  </li>
                  <li>
                    <Link href="http://www.formette.com/docs" target="_blank">
                      Help
                    </Link>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const profileWithData = compose(
  graphql(USER_QUERY, { name: "userQuery" }),
  graphql(UPDATE_USER_MUTATION, { name: "updateUser" })
)(Profile);

export default withApollo(profileWithData);
