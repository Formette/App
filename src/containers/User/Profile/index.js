// @flow
import React, { PureComponent } from "react";
import { graphql, compose, withApollo } from "react-apollo";
//Components
import {
  SubTitle,
  Input,
  Button,
  Text,
  Link,
  Header
} from "../../../components/atoms";
import { Confirmation, Card } from "../../../components/molecules";
//hocs
import { withUser } from "../../../hocs";
//Utils
import { _logout, _formatUsername } from "../../../services/utilities";
import LogRocket from "logrocket";
import { withAlert } from "react-alert";
//API
import { USERNAME_VALIDATION_QUERY } from "../../../api/Queries";
import { UPDATE_USER_MUTATION } from "../../../api/Mutations";

export class Profile extends PureComponent {
  props: {
    updateUser: any,
    client: any,
    router: any
  };
  state = {
    username: "",
    error: false,
    errorMsg: "",
    timeoutUserName: 0,
    onConfirmation: false
  };
  _showConfirmation = () => {
    this.setState(prevState => ({
      onConfirmation: !prevState.onConfirmation
    }));
    LogRocket.track("Opened  modal on change username");
  };
  _updateProfile = async () => {
    const { error } = this.state;
    let username = _formatUsername(this.state.username);
    //hides the confirmation modal
    this.setState(prevState => ({
      onConfirmation: !prevState.onConfirmation
    }));
    //checks is the username is the same as the previous one
    if (this._isTheSameUsername(username)) return;
    //Verifies if the inputs are empty or not
    if (username) {
      if (error) {
        return;
      }
      const userId = this.props.user.state.profile.id;
      //updates the user username and some else info in the DB
      try {
        await this.props.updateUser({
          variables: {
            userId,
            username
          }
        });
        //Shows feedback and updates the localStorage
        this.setState({ username });
        this.props.user.changeUsername(username);
        this.props.alert.success("Change made successfully");
        LogRocket.info("Change made successfully");
        LogRocket.track("Updated username");
      } catch (e) {
        LogRocket.error({ _updateProfile: e });
        this.props.alert.error("Something went wrong, try again...");
      }
    } else {
      LogRocket.warn(
        "This form is feeling lonely, needs affection, needs data."
      );
      this.setState({
        error: true,
        errorMsg: "This form is feeling lonely, needs affection, needs data."
      });
    }
  };
  _onUsernameValidation(getUsername: string) {
    clearTimeout(this.state.timeoutUserName);
    this.setState({
      timeoutUserName: setTimeout(() => {
        //checks is the username is the same as the previous one
        let username = _formatUsername(getUsername);
        if (this._isTheSameUsername(username)) return;
        this.props.client
          .query({
            query: USERNAME_VALIDATION_QUERY,
            variables: { username }
          })
          .then(res => {
            if (Object.keys(res.data.allUsers).length !== 0) {
              LogRocket.warn(
                "With so much name in this world, you had to choose this one. Try another."
              );
              this.setState({
                error: true,
                errorMsg:
                  "With so much name in this world, you had to choose this one. Try another."
              });
            } else {
              this.setState({ error: false });
            }
          })
          .catch(e => {
            LogRocket.error({ _onUsernameValidation: e });
          });
      }, 500)
    });
  }
  _isTheSameUsername(username: string) {
    const { user } = this.props;
    if (user.state.userName === username) {
      LogRocket.warn(
        "If it's the same as before, what's the point of changing?"
      );
      this.setState({
        error: true,
        errorMsg: "If it's the same as before, what's the point of changing?"
      });
      return true;
    }
  }
  render() {
    const { error, errorMsg, onConfirmation } = this.state;
    const { userName, formsesMeta } = this.props.user.state;
    return (
      <div>
        <Confirmation
          title="Are you sure?"
          description="All your endpoints will be changed to the new username. Do not forget to change in your apps."
          show={onConfirmation}
          onCancel={this._showConfirmation}
          onConfirmation={this._updateProfile}
          onConfirmationText="Confirm"
          actionProps={{ primary: true }}
        />
        <div className="row">
          <div className="col-md-12">
            <Header>{`Hey, ${this.state.username || userName}`}</Header>
          </div>
        </div>
        <div className="row">
          <div className="col-md-6">
            <form>
              <div className={`form-group ${error ? "has-danger" : ""}`}>
                <Text highlight>Username:</Text>
                <Input
                  placeholder="username"
                  defaultValue={userName}
                  onKeyUp={e => this._onUsernameValidation(e.target.value)}
                  onChange={e =>
                    this.setState({ username: e.target.value, error: false })
                  }
                  className="form-control"
                />
                {error ? <Text color="red">{errorMsg}</Text> : ""}
              </div>
              <Button
                onClick={this._showConfirmation}
                className="btn btn-lg btn-primary btn-block"
                primary
              >
                Update profile
              </Button>
            </form>
          </div>
          <div className="col-md-6">
            <Card style={{ marginTop: 10 }}>
              <div className="card-body">
                <Text highlight>Statistics:</Text>
                <Text>
                  We are great analysts, here you have your statistics of your
                  forms.
                </Text>
                <SubTitle>{`${formsesMeta} forms created`}</SubTitle>
              </div>
            </Card>
            <Card style={{ marginTop: 10 }}>
              <div className="card-body">
                <Text highlight>Current Plan:</Text>
                <Text>
                  More plans soon, as the platform is in beta you have access to
                  all features.
                </Text>
                <SubTitle>
                  <span className="badge badge-dark">All the features</span>
                </SubTitle>
              </div>
            </Card>
            <Card style={{ marginTop: 10 }}>
              <div className="card-body">
                <Text highlight>Settings:</Text>
                <Text>
                  This is a dangerous zone, so be careful, here you will find
                  your settings.
                </Text>
                <ul className="list-inline">
                  <li>
                    <Link onClick={_logout}>Log Out</Link>
                  </li>
                  <li>
                    <Link href="http://www.formette.com/docs" target="_blank">
                      Help
                    </Link>
                  </li>
                </ul>
              </div>
            </Card>
          </div>
        </div>
      </div>
    );
  }
}

const profileWithData = compose(
  withUser,
  withAlert,
  graphql(UPDATE_USER_MUTATION, { name: "updateUser" })
)(Profile);

export default withApollo(profileWithData);
