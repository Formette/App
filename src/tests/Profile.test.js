// Profile.test.js
//Utils
import "./helpers/browser";
import "./helpers/mock-localstorage";
import React from "react";
import { Profile } from "../containers/Profile";
import renderer from "react-test-renderer";

describe("Profile", () => {
  it("UI renders correctly", () => {
    const data = {
      user: {
        email: "amaralvitor@ua.pt",
        id: "cj5frzfv11z6y0180k67lraep",
        userName: "vacom",
        _formsesMeta: {
          count: 3
        }
      }
    };

    const tree = renderer.create(<Profile userQuery={data} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
