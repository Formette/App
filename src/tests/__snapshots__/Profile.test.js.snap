// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile UI renders correctly 1`] = `
<div>
  <div
    className="css-1bdcmy7"
  >
    <span />
  </div>
  <div
    className="card mx-auto animated zoomInUp sc-kAzzGY cfOeAw"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div
      className="card-body"
    >
      <h3
        className="sc-ifAKCX dYpvlb"
      >
        Are you sure?
      </h3>
      <p
        className="sc-EHOje fkFstZ"
      >
        All your endpoints will be changed to the new username. Do not forget to change in your apps.
      </p>
      <div
        className="text-right"
      >
        <a
          className="cancelAction sc-htpNat jfadgg"
          onClick={[Function]}
        >
          Cancel
        </a>
        <button
          className="btn sc-bdVaJa fFweCv"
          color="#12b886"
          onClick={[Function]}
          type="button"
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      <h2
        className="sc-bxivhb kRbHZm"
      >
        Hey, undefined
      </h2>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <form>
        <div
          className="form-group "
        >
          <h6
            className="sc-bZQynM ekLzcz"
          >
            Username:
          </h6>
          <input
            className="form-control sc-bwzfXH kUMhZx"
            defaultValue="vacom"
            onChange={[Function]}
            onKeyUp={[Function]}
            placeholder="username"
            type="text"
          />
          
        </div>
        <button
          className="btn btn-lg btn-primary btn-block sc-bdVaJa hayiWv"
          color="#7950f2"
          onClick={[Function]}
          type="button"
        >
          Update profile
        </button>
      </form>
    </div>
    <div
      className="col-md-6"
    >
      <div
        className="card"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <div
          className="card-body"
        >
          <h6
            className="sc-bZQynM ekLzcz"
          >
            Statistics:
          </h6>
          <p
            className="sc-EHOje VGcDz"
          >
            We are great analysts, here you have your statistics of your forms.
          </p>
          <h3
            className="sc-ifAKCX dYpvlb"
          >
            3 forms created
          </h3>
        </div>
      </div>
      <div
        className="card"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <div
          className="card-body"
        >
          <h6
            className="sc-bZQynM ekLzcz"
          >
            Current Plan:
          </h6>
          <p
            className="sc-EHOje VGcDz"
          >
            More plans soon, as the platform is in beta you have access to all features.
          </p>
          <h2>
            <span
              className="badge badge-dark"
            >
              All the features
            </span>
          </h2>
        </div>
      </div>
      <div
        className="card"
        style={
          Object {
            "marginTop": 10,
          }
        }
      >
        <div
          className="card-body"
        >
          <h6
            className="sc-bZQynM ekLzcz"
          >
            Settings:
          </h6>
          <p
            className="sc-EHOje VGcDz"
          >
            This is a dangerous zone, so be careful, here you will find your settings.
          </p>
          <ul
            className="list-inline"
          >
            <li>
              <a
                className="sc-htpNat jfadgg"
                onClick={[Function]}
              >
                Log Out
              </a>
            </li>
            <li>
              <a
                className="sc-htpNat jfadgg"
                href="http://www.formette.com/docs"
                target="_blank"
              >
                Help
              </a>
            </li>
            <li>
              <a
                className="sc-htpNat iZrbGc"
                color="#fa5252"
              >
                Delete Account
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
